File: priv/lance_js/src/schema.js
const arrow = require("apache-arrow");

function cfgToField(cfg) {
  const field = function (name, type, nullable) {
    return [new arrow.Field(name, type, nullable)];
  };
  const { name, field_type, nullable } = cfg;
  switch (field_type.type) {
    case "utf8":
      return field(name, new arrow.Utf8(), nullable);
    case "float":
      return field(name, new arrow.Float(field_type.precision), nullable);
    case "int32":
      return field(name, new arrow.Int32(), nullable);
    case "float32":
      return field(name, new arrow.Float32(), nullable);
    case "list":
      return [
        new arrow.Field(
          name,
          new arrow.List(cfgToField(field_type.child)[0]),
          nullable
        ),
      ];
    case "fixed_size_list":
      return [
        new arrow.Field(
          name,
          new arrow.FixedSizeList(
            field_type.dimension,
            cfgToField(field_type.child)[0]
          ),
          nullable
        ),
      ];

    default:
      return [];
  }
}

function fieldConfigsToSchema(fieldConfigs) {
  const fields = fieldConfigs.map(cfgToField).flat();
  return new arrow.Schema(fields);
}

module.exports = {
  fieldConfigsToSchema,
};

File: native/elixir_lancedb/src/schema.rs
use arrow_schema::{DataType, Field as ArrowField};
use rustler::{NifStruct, NifTaggedEnum, NifUnitEnum};
use std::{collections::HashMap, sync::Arc};

#[derive(NifStruct, Clone)]
#[module = "ElixirLanceDB.Native.Schema.Field"]
pub struct Field {
    pub name: String,
    pub field_type: FieldType,
    pub nullable: bool,
}

impl Field {
    fn into_arrow(self) -> ArrowField {
        match self.field_type {
            FieldType::Utf8 => ArrowField::new(self.name, DataType::Utf8, self.nullable),
            FieldType::Float32 => ArrowField::new(self.name, DataType::Float32, self.nullable),
            FieldType::Int32 => ArrowField::new(self.name, DataType::Int32, self.nullable),
            FieldType::List(child) => ArrowField::new(
                self.name,
                DataType::List(Arc::new(child.into_arrow())),
                self.nullable,
            ),
            FieldType::FixedSizeList(child, dimension) => ArrowField::new(
                self.name,
                DataType::FixedSizeList(Arc::new(child.into_arrow()), dimension),
                self.nullable,
            ),
        }
    }
}

#[derive(NifStruct, Clone, Debug)]
#[module = "ElixirLanceDB.Native.Schema.Field"]
pub struct ChildField {
    pub name: String,
    pub field_type: ChildFieldType,
    pub nullable: bool,
}

impl ChildField {
    fn into_arrow(self) -> ArrowField {
        match self.field_type {
            ChildFieldType::Utf8 => ArrowField::new(self.name, DataType::Utf8, self.nullable),
            ChildFieldType::Float32 => ArrowField::new(self.name, DataType::Float32, self.nullable),
            ChildFieldType::Int32 => ArrowField::new(self.name, DataType::Int32, self.nullable),
        }
    }
}

#[derive(NifTaggedEnum, Debug, Clone)]
pub enum FieldType {
    Utf8,
    Float32,
    Int32,
    List(ChildField),
    FixedSizeList(ChildField, i32),
}

#[derive(NifUnitEnum, Clone, Copy, Debug)]
pub enum ChildFieldType {
    Utf8,
    Float32,
    Int32,
}

#[derive(NifStruct, Clone)]
#[module = "ElixirLanceDB.Native.Schema"]
pub struct Schema {
    pub fields: Vec<Field>,
    pub metadata: HashMap<String, String>,
}

impl Schema {
    pub fn into_arrow(self) -> arrow_schema::Schema {
        let fields: Vec<ArrowField> = self.fields.iter().map(|f| f.clone().into_arrow()).collect();
        arrow_schema::Schema::new(fields)
    }
}
